contracts/src/batch-bridge/L1BatchBridgeGateway.sol

```diff
@@ -0,0 +1,425 @@
+    }
+
+    /*****************************
+     * Public Mutating Functions *
+     *****************************/
+
+    /// @notice Receive refunded ETH from `L1ScrollMessenger`.
+    receive() external payable {
+        if (_msgSender() != messenger) {
+            revert ErrorCallerNotMessenger();
+        }
+    }
+
+    /// @notice Deposit ETH.
+    function depositETH() external payable {
+        // no safe cast check here, since no one has so much ETH yet.
+        _deposit(address(0), _msgSender(), uint96(msg.value));
+    }
+
+    /// @notice Deposit ERC20 token.
+    ///
+    /// @param token The address of token.
+    /// @param amount The amount of token to deposit. We use type `uint96`, since it is enough for most of the major tokens.
+    function depositERC20(address token, uint96 amount) external {
+        if (token == address(0)) revert ErrorIncorrectMethodForETHDeposit();
+
+        // common practice to handle fee on transfer token.
+        uint256 beforeBalance = IERC20Upgradeable(token).balanceOf(address(this));
+        IERC20Upgradeable(token).safeTransferFrom(_msgSender(), address(this), amount);
+        amount = uint96(IERC20Upgradeable(token).balanceOf(address(this)) - beforeBalance);
+
+        _deposit(token, _msgSender(), amount);
+    }
+
+    /************************
+     * Restricted Functions *
+     ************************/
+
+    /// @notice Add or update the batch bridge config for the given token.
+    ///
+    /// @dev The caller should make sure `safeBridgeGasLimit` is enough for batch bridging.
+    ///
+    /// @param token The address of token to update.
+    /// @param newConfig The new config.
+    function setBatchConfig(address token, BatchConfig memory newConfig) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (
+            newConfig.maxTxsPerBatch == 0 ||
+            newConfig.maxDelayPerBatch == 0 ||
+            newConfig.feeAmountPerTx > newConfig.minAmountPerTx
+        ) {
+            revert ErrorInvalidBatchConfig();
+        }
+        configs[token] = newConfig;
+    }
+
+    /// @notice Initiate the batch bridge of current pending batch.
+    /// @param token The address of the token.
+    function executeBatchDeposit(address token) external payable onlyRole(KEEPER_ROLE) {
+        BatchConfig memory cachedBatchConfig = configs[token];
+        TokenState memory cachedTokenState = tokens[token];
+        _tryFinalizeCurrentBatch(token, cachedBatchConfig, cachedTokenState);
+
+        // no batch to bridge
+        if (cachedTokenState.currentBatchIndex == cachedTokenState.pendingBatchIndex) {
+            revert ErrorNoPendingBatch();
+        }
+
+        // check bridge fee
+        uint256 depositFee = IL1MessageQueue(queue).estimateCrossDomainMessageFee(cachedBatchConfig.safeBridgeGasLimit);
+        uint256 batchBridgeFee = IL1MessageQueue(queue).estimateCrossDomainMessageFee(SAFE_BATCH_BRIDGE_GAS_LIMIT);
+        if (msg.value < depositFee + batchBridgeFee) {
```

Comment:
Just to understand the two fees. The batch bridge fee is the gas for execution in the scroll messenger contract while the deposit fee is the gas for ETH or ERC20 transfers. Is this right?

Reply:
That is how I understand it, yes.
cc @zimpha

Reply:
yeah, the `depositFee` is for bridge ETH/ERC20 from L1 to L2 and the `batchBridgeFee` is for notifying the deposit to L2.